surprise-pairwise-from-embedding.R

```diff
@@ -0,0 +1,151 @@
+library(lsa)
+library(tokenizers)
+library(dplyr)
+library(gtools) #pairwise combinations
+#load tsv
+drama_embedding <- read_tsv("YOUR_DIR/hamlet_embeddings.tsv")
+
+#prerocess the data: 
+# - embeddings as numbers
+# - instead of timestamp just count the nth position of a sentence
+drama_embedding <- drama_embedding %>%
+  rowwise() %>%
+  mutate(embedding_num = list(reticulate::py_eval(embeddings)))
+drama_embedding$created_at <- 1:nrow(drama_embedding)
+
+#save embedding values as numeric in a separate list 
+#- we will measure the cosine similarity between the elements of this list
+a1 <- lapply(drama_embedding$embedding_num, as.numeric)
+# keep the sentences with more than 4 words
+tok_sentence <- tokenize_words(drama_embedding$speech_sent)
+tok_sentence_length <- sapply(tok_sentence, length)
+longsent <- which(tok_sentence_length > 4)
+drama_longsent <- drama_embedding[drama_embedding$created_at %in% longsent, ]
+
+
+# just if not filtered the characters with more than 1000 words beforhand
+drama_short <- drama_longsent %>%
+  filter( character == "#Claudius_Ham" | character == "#Gertrude_Ham" | 
+            character =="#Hamlet_Ham" | character == "#Horatio_Ham" | 
+            character == "#Laertes_Ham" | character == "#Ophelia_Ham" | 
+            character == "#Polonius_Ham")
+
+# castlist
+cast <- sort(unique(drama_short$character))
+
+#  sentences of a character in separate tibble
+chars <- drama_short %>% 
+  group_by(character)
+chars <- group_split(chars)
+names(chars) <- cast
+
+# get all the possible pairwise combinations of char names
+pairs <-as.data.frame(permutations(length(cast), 2, cast, repeats = FALSE))
+colnames(pairs) <- c("source", "target")
+
+## main calculation of the semantically closest sentences between S(ource) and T(arget)
+
+#list for the tibbles of the pairwise measures
+pair_sim <- list()
+# loop iterates over the pairs (firs column = source, second = target)
+for(k in 1:nrow(pairs)){
+  source <- chars[[which(names(chars) == pairs[k,1])]]
+  target <- chars[[which(names(chars) == pairs[k,2])]]
+  
+  all_sim <- list() #list containing all cosine similarity between target sentence and previous source sentences
+  max_sim <- list() #list containing maximum cosine similarity between a target sentence and previous source sentences
+  
+  #loop ove the target sentences
+  for(j in 1:nrow(target)){
+    #loop over source sentences
+    for(i in 1:nrow(source)){
+      if(source[i,2]<target[j,2]){
+        all_sim[[i]] <-cosine(a1[[as.numeric(target[j,2])]], 
+                              a1[[as.numeric(source[i,2])]])
+      }
+      else{ #if there is no previous sentences
+        all_sim[[i]] <- 0}

```

Comment:
Note that the minimum cosine similarity is -1, not 0, so in theory this might be the highest value in the set. In practice, I can't imagine it ever would be.